File: backend/cmd/main/gamestate_binary.go
package main

import (
	"bytes"
	"encoding/binary"
)

// decodeGameStateBinary декодирует бинарный формат в GameState
//
//lint:ignore U1000 Используется для отладки и тестирования
func decodeGameStateBinary(data []byte) (*GameState, error) {
	buf := bytes.NewReader(data)

	gs := &GameState{}

	// Читаем размеры
	var rows, cols, mines, revealed uint16
	binary.Read(buf, binary.LittleEndian, &rows)
	binary.Read(buf, binary.LittleEndian, &cols)
	binary.Read(buf, binary.LittleEndian, &mines)
	binary.Read(buf, binary.LittleEndian, &revealed)

	gs.Rows = int(rows)
	gs.Cols = int(cols)
	gs.Mines = int(mines)
	gs.Revealed = int(revealed)

	// Читаем HintsUsed
	hintsUsed, _ := buf.ReadByte()
	gs.HintsUsed = int(hintsUsed)

	// Читаем флаги
	flags, _ := buf.ReadByte()
	gs.GameOver = (flags & (1 << 0)) != 0
	gs.GameWon = (flags & (1 << 1)) != 0

	// Читаем LoserPlayerID
	loserPIDLen, _ := buf.ReadByte()
	pidBytes := make([]byte, 5)
	buf.Read(pidBytes)
	if loserPIDLen > 0 && loserPIDLen <= 5 {
		gs.LoserPlayerID = string(pidBytes[:loserPIDLen])
	}

	// Читаем LoserNickname
	nicknameLen, _ := buf.ReadByte()
	if nicknameLen > 0 {
		nicknameBytes := make([]byte, nicknameLen)
		buf.Read(nicknameBytes)
		gs.LoserNickname = string(nicknameBytes)
	}

	// Инициализируем Board
	gs.Board = make([][]Cell, gs.Rows)
	for i := range gs.Board {
		gs.Board[i] = make([]Cell, gs.Cols)
	}

	// Читаем Board
	for i := 0; i < gs.Rows; i++ {
		for j := 0; j < gs.Cols; j++ {
			cellByte, _ := buf.ReadByte()
			cell := &gs.Board[i][j]
			cell.IsMine = (cellByte & (1 << 0)) != 0
			cell.IsRevealed = (cellByte & (1 << 1)) != 0
			cell.IsFlagged = (cellByte & (1 << 2)) != 0
			cell.NeighborMines = int((cellByte >> 3) & 0x0F) // Бит 3-6
		}
	}

	// Читаем цвета флагов (если есть данные)
	if buf.Len() > 0 {
		flagCount, err := buf.ReadByte()
		if err == nil && flagCount > 0 {
			// Читаем цвета флагов
			for i := 0; i < int(flagCount); i++ {
				// Читаем cellKey (2 байта)
				var cellKey uint16
				if err := binary.Read(buf, binary.LittleEndian, &cellKey); err != nil {
					break
				}
				// Читаем длину цвета
				colorLen, err := buf.ReadByte()
				if err != nil {
					break
				}
				// Читаем цвет
				if colorLen > 0 && colorLen <= 7 {
					colorBytes := make([]byte, colorLen)
					if n, err := buf.Read(colorBytes); err == nil && n == int(colorLen) {
						color := string(colorBytes)
						// Применяем цвет к соответствующей ячейке
						row := int(cellKey) / gs.Cols
						col := int(cellKey) % gs.Cols
						if row >= 0 && row < gs.Rows && col >= 0 && col < gs.Cols {
							gs.Board[row][col].FlagColor = color
						}
					}
				}
			}
		}
	}

	// Читаем SafeCells (если есть данные)
	if buf.Len() > 0 {
		var safeCellsCount uint16
		if err := binary.Read(buf, binary.LittleEndian, &safeCellsCount); err == nil && safeCellsCount > 0 {
			gs.SafeCells = make([]SafeCell, safeCellsCount)
			for i := 0; i < int(safeCellsCount); i++ {
				var row, col uint16
				if err := binary.Read(buf, binary.LittleEndian, &row); err != nil {
					break
				}
				if err := binary.Read(buf, binary.LittleEndian, &col); err != nil {
					break
				}
				gs.SafeCells[i] = SafeCell{Row: int(row), Col: int(col)}
			}
		}
	}

	// Читаем CellHints (если есть данные)
	if buf.Len() > 0 {
		var hintsCount uint16
		if err := binary.Read(buf, binary.LittleEndian, &hintsCount); err == nil && hintsCount > 0 {
			gs.CellHints = make([]CellHint, hintsCount)
			for i := 0; i < int(hintsCount); i++ {
				var row, col uint16
				if err := binary.Read(buf, binary.LittleEndian, &row); err != nil {
					break
				}
				if err := binary.Read(buf, binary.LittleEndian, &col); err != nil {
					break
				}
				hintTypeByte, err := buf.ReadByte()
				if err != nil {
					break
				}
				var hintType string
				switch hintTypeByte {
				case 0:
					hintType = "MINE"
				case 1:
					hintType = "SAFE"
				case 2:
					hintType = "UNKNOWN"
				default:
					hintType = "UNKNOWN"
				}
				gs.CellHints[i] = CellHint{Row: int(row), Col: int(col), Type: hintType}
			}
		}
	}

	return gs, nil
}

File: frontend/src/utils/gamestateBinary.ts
/**
 * Декодирует бинарный формат gameState
 * Формат:
 * - 2 байта: Rows (uint16, little-endian)
 * - 2 байта: Cols (uint16, little-endian)
 * - 2 байта: Mines (uint16, little-endian)
 * - 2 байта: Revealed (uint16, little-endian)
 * - 1 байт: HintsUsed (количество использованных подсказок, 0-3)
 * - 1 байт: Флаги (бит 0: GameOver, бит 1: GameWon)
 * - 1 байт: Длина LoserPlayerID (0-5)
 * - 5 байт: LoserPlayerID (ASCII)
 * - 1 байт: Длина LoserNickname
 * - N байт: LoserNickname (UTF-8)
 * - Rows*Cols байт: Board (каждая Cell = 1 байт)
 * - 1 байт: Количество флагов с цветами
 * - Для каждого флага: 2 байта cellKey (uint16), 1 байт длина цвета, N байт цвет
 * - 2 байта: Количество SafeCells (uint16)
 * - Для каждой SafeCell: 2 байта row (uint16), 2 байта col (uint16)
 */
export function decodeGameStateBinary(data: ArrayBuffer): {
  b: Array<Array<{ m: boolean; r: boolean; f: boolean; n: number; fc?: string }>> // board
  r: number // rows
  c: number // cols
  m: number // mines
  go: boolean // gameOver
  gw: boolean // gameWon
  rv: number // revealed
  hu: number // hintsUsed
  sc?: Array<{ r: number; c: number }> // safeCells
  hints?: Array<{ r: number; c: number; t: string }> // cellHints - подсказки для ячеек (MINE, SAFE, UNKNOWN)
  lpid?: string // loserPlayerId
  ln?: string // loserNickname
} {
  const view = new DataView(data)
  let offset = 0

  // Читаем размеры (uint16, little-endian)
  const rows = view.getUint16(offset, true)
  offset += 2
  const cols = view.getUint16(offset, true)
  offset += 2
  const mines = view.getUint16(offset, true)
  offset += 2
  const revealed = view.getUint16(offset, true)
  offset += 2

  // Читаем HintsUsed
  const hintsUsed = view.getUint8(offset)
  offset += 1

  // Читаем флаги
  const flags = view.getUint8(offset)
  offset += 1
  const gameOver = (flags & (1 << 0)) !== 0
  const gameWon = (flags & (1 << 1)) !== 0

  // Читаем LoserPlayerID
  const loserPIDLen = view.getUint8(offset)
  offset += 1
  const pidBytes = new Uint8Array(data, offset, 5)
  offset += 5
  let loserPlayerID: string | undefined
  if (loserPIDLen > 0 && loserPIDLen <= 5) {
    loserPlayerID = new TextDecoder('ascii').decode(pidBytes.slice(0, loserPIDLen))
  }

  // Читаем LoserNickname
  const nicknameLen = view.getUint8(offset)
  offset += 1
  let loserNickname: string | undefined
  if (nicknameLen > 0) {
    const nicknameBytes = new Uint8Array(data, offset, nicknameLen)
    offset += nicknameLen
    loserNickname = new TextDecoder('utf-8').decode(nicknameBytes)
  }

  // Инициализируем Board
  const board: Array<Array<{ m: boolean; r: boolean; f: boolean; n: number; fc?: string }>> = []
  for (let i = 0; i < rows; i++) {
    board[i] = []
    for (let j = 0; j < cols; j++) {
      const cellByte = view.getUint8(offset)
      offset += 1
      
      const cell = {
        m: (cellByte & (1 << 0)) !== 0, // IsMine
        r: (cellByte & (1 << 1)) !== 0, // IsRevealed
        f: (cellByte & (1 << 2)) !== 0, // IsFlagged
        n: (cellByte >> 3) & 0x0F // NeighborMines (бит 3-6)
      }
      board[i][j] = cell
    }
  }

  // Читаем цвета флагов (если есть данные)
  if (offset < data.byteLength) {
    const flagCount = view.getUint8(offset)
    offset += 1
    
    // Читаем цвета флагов
    for (let i = 0; i < flagCount && offset < data.byteLength; i++) {
      // Читаем cellKey (2 байта, little-endian)
      const cellKey = view.getUint16(offset, true)
      offset += 2
      
      // Читаем длину цвета
      if (offset >= data.byteLength) break
      const colorLen = view.getUint8(offset)
      offset += 1
      
      // Читаем цвет
      if (colorLen > 0 && colorLen <= 7 && offset + colorLen <= data.byteLength) {
        const colorBytes = new Uint8Array(data, offset, colorLen)
        offset += colorLen
        const color = new TextDecoder('utf-8').decode(colorBytes)
        
        // Применяем цвет к соответствующей ячейке
        const row = Math.floor(cellKey / cols)
        const col = cellKey % cols
        if (row >= 0 && row < rows && col >= 0 && col < cols) {
          board[row][col].fc = color
        }
      }
    }
  }

  // Читаем SafeCells (если есть данные)
  let safeCells: Array<{ r: number; c: number }> | undefined
  if (offset < data.byteLength) {
    const safeCellsCount = view.getUint16(offset, true)
    offset += 2
    
    if (safeCellsCount > 0) {
      safeCells = []
      for (let i = 0; i < safeCellsCount && offset + 4 <= data.byteLength; i++) {
        const row = view.getUint16(offset, true)
        offset += 2
        const col = view.getUint16(offset, true)
        offset += 2
        safeCells.push({ r: row, c: col })
      }
    }
  }

  return {
    b: board,
    r: rows,
    c: cols,
    m: mines,
    go: gameOver,
    gw: gameWon,
    rv: revealed,
    hu: hintsUsed,
    sc: safeCells,
    lpid: loserPlayerID,
    ln: loserNickname
  }
}

